doc/sphinxext/missing_references.py

```diff
@@ -0,0 +1,241 @@
+3. Run sphinx-build. It will generate ``missing-references.json``
+    next to your ``conf.py``.
+4. Remove ``missing_references_write_json = True`` from your
+    ``conf.py`` (or set it to ``False``)
+5. Run sphinx-build again, and ``nitpick_ignore`` will
+    contain all of the previously failed references.
+
+"""
+
+from collections import defaultdict
+import json
+import logging
+import os.path
+import posixpath
+
+from docutils.utils import get_source_line
+from docutils import nodes
+from sphinx.util import logging as sphinx_logging
+
+logger = sphinx_logging.getLogger(__name__)
+
+
+class MissingReferenceFilter(logging.Filter):
+    """
+    A logging filter designed to record missing reference warning messages
+    for use by this extension
+    """
+    def __init__(self, app):
+        self.app = app
+        super().__init__()
+
+    def _record_reference(self, record):
+        if not (getattr(record, 'type', '') == 'ref' and
+                isinstance(getattr(record, 'location', None), nodes.Node)):
+            return
+
+        if not hasattr(self.app.env, "missing_reference_warnings"):
+            self.app.env.missing_reference_warnings = defaultdict(set)
+
+        record_missing_reference(self.app,
+            self.app.env.missing_reference_warnings,
+            record.location)
+
+    def filter(self, record):
+        self._record_reference(record)
+        return True
+
+
+def record_missing_reference(app, record, node):
+    domain = node["refdomain"]
+    typ = node["reftype"]
+    target = node["reftarget"]
+    location = get_location(node, app)
+
+    dtype = "{}:{}".format(domain, typ)
+
+    record[(dtype, target)].add(location)
+
+
+def record_missing_reference_handler(app, env, node, contnode):
+    """
+    When the sphinx app notices a missing reference, it emits an
+    event which calls this function. This function records the missing
+    references for analysis at the end of the sphinx build.
+    """
+    if not app.config.missing_references_enabled:
+        # no-op when we are disabled.
+        return
+
+    if not hasattr(env, "missing_reference_events"):
+        env.missing_reference_events = defaultdict(set)
+
+    record_missing_reference(app, env.missing_reference_events, node)
+
+
+def get_location(node, app):
+    """
+    Given a docutils node and a sphinx application, return a string
+    representation of the source location of this node.
+
+    Usually, this will be of the form "path/to/file:linenumber". Two
+    special values can be emitted, "<external>" for paths which are
+    not contained in this source tree (e.g. docstrings included from
+    other modules) or "<unknown>", inidcating that the sphinx application
+    cannot locate the original source file (usually because an extension
+    has injected text into the sphinx parsing engine).
+    """
+    (path, line) = get_source_line(node)
+
+    if path:
+
+        basepath = os.path.abspath(os.path.join(app.confdir, ".."))
```

Comment:
actually I think this will fail if building the docs against a non-editable install? (so that you'd find the wrong matplotlib by going just up once from doc/conf.py)  Admittedly a bit theoretical, but you could possibly instead normalize the paths relative to `Path(matplotlib.__init__.__file__).parent.parent` (which helpfully also works for mpl-toolkits) instead (I guess if this was to be a mpl-independent extension you'd need to make this a config option, but note that in such a case your current approach also fails if conf.py is under doc/source/conf.py rather than doc/conf.py, which is a valid sphinx layout).

Reply:
Ahh – yes, this makes sense. A very good catch.

I'm going to test your suggestion and see what happens. It definitely won't work if this becomes a true sphinx extension, but there are are a few ways around that in the future.

Reply:
I've tried some things out here –

It's quite difficult to gracefully handle building the docs correctly when `matplotlib` isn't an editable install. Figuring out what is a "matplotlib" path vs any other module, and producing paths which match the editable install case, is very difficult. e.g. in the source tree mode, we have a path `lib/matplotlib/axes/_axes.py`, but in installed mode, this is something like `site-packages/matplotlib/axes/_axes.py`. Although we could normalize those two paths, it is a lot of work for a particular edge case, for functionality that isn't very important to this extension. (The features of this extension are already ballooning beyond what I had envisioned).

Instead, I propose that we detect when the matplotlib source directory doesn't seem to be in the right place _relative to the docs directory_ and in those cases, we don't worry about warning for unused ignores. Everything else should still work fine in those cases – the actual process of ignoring unmatched references doesn't use the file paths at all.

Reply:
Sorry to have sent you on a wild goose chase... just ignoring warnings if not in a standard layout is perfectly fine.

Reply:
No worries – it was interesting idea to try. I'm just trying to balance making things useful with making the perfect sphinx extension!